src/diagram/diagram-api.tsx

```diff
@@ -127,25 +223,53 @@ export type Edge = {
 
 }
 
+
 /**
  * Callbacks to owner to handle required user actions.
  */
 export interface DiagramCallbacks {
 
+
   // Node
 
+  /**
+   * This property stores the method, which is called when user opens node's detail.
+   * @param identifier is the identifier of the node for which the detail was shown.
+   */
   onShowNodeDetail: (identifier: string) => void;
 
+
+  /**
+   * This property stores the method, which is called when user starts editing node.
+   * @param identifier is the identifier of the node which is being edited.
+   */
   onEditNode: (identifier: string) => void;
 
+
+  /**
+   * This property stores the method, which is called when user starts creating node's profile.
+   * @param identifier is the identifier of the node of which the profile is being created.
+   */
   onCreateNodeProfile: (identifier: string) => void;
 
+
+  /**
+   * This property stores the method, which is called when user hides node, i. e. removes it from canvas.
+   * @param identifier is the identifier of the node, which is newly hidden.
+   */
   onHideNode: (identifier: string) => void;
 
+
+  /**
+   * This property stores the method, which is called when user deletes node.
+   * @param identifier is the identifier of the deleted node.
+   */
   onDeleteNode: (identifier: string) => void;
```

Comment:
I was thinking about the API functionality here and I see difference between some callbacks (actually in the end I came to conclusion that not necesarily), for example `onDeleteNode` and `onEditNode`. 

Right now I understand it that the flow for deletion works this way.

1. User clicks 🗑️
2. The editor removes node from canvas
3. Calls the `onDeleteNode` callback method which removes it from the semantic model.

But in case of the `onEditNode` the flow is the following:

1. User clicked `Edit class` in context menu
2. The `onEditNode` callback is called. In the callback probably edit dialog is opened through dialog API.
3. The callback has to notify editor back with the change.

So in one case the callback is something which is called after editor's action, in the other case it is something, which actually performs the action and notifies the editor back with change, so it can react accordingly. 
But I guess that in the `onDeleteNode` case we can just call the callback and then notify back after the actual deletion in the semantic model that it no longer exists and the editor should remove it. (so basically switch 2. and 3.)

Reply:
They way I think about it the onDeleteNode actually call the owner who is responsible for removing the node. In other word, the editor does not remove the node on it's own. 

We can also think whether waypoints should also be added by the owner and not the editor. Any ideas on this?

Reply:
So just to make sure I understand it correctly. The idea is:

1. User clicks 🗑️
2. `onDeleteNode` is called
3. The owner (Something which created the editor - meaning it created the instance and passed it implementations of actions) removes the node from semantic model.
4. After it is removed depending on the implementation one of the two following things happen
- Directly after that `removeNodes` on `DiagramActions` is called within the `onDeleteNode` callback
- There is registered callback (from the editor component?) on visual model, which removes node in the editor component if it was removed in the visual model. I guess that it is how it is implemented in cme-v1.

**Waypoints:**
If it should be added by the owner then the flow is the following?

1. User clicks `+` in the editor to add waypoint.
2. `onCreateWaypoint` or something like that is called.
3. Here the owner, which has access to the visual model updates the visual model.
4. Similar to the removal.

I guess that it makes sense based on my basic understanding.